src/crimson/os/seastore/extent_placement_manager.cc

```diff
@@ -424,20 +454,28 @@ seastar::future<>
 ExtentPlacementManager::BackgroundProcess::do_background_cycle()
 {
   assert(is_ready());
-  if (trimmer->should_trim_alloc()) {
+  if (trimmer->should_trim_alloc() &&
+      cleaner->try_reserve_projected_usage(
+          trimmer->get_trim_alloc_size_per_cycle())) {
     return trimmer->trim_alloc(
     ).handle_error(
       crimson::ct_error::assert_all{
 	"do_background_cycle encountered invalid error in trim_alloc"
       }
-    );
-  } else if (trimmer->should_trim_dirty()) {
+    ).finally([this]{
+      cleaner->release_projected_usage(trimmer->get_trim_alloc_size_per_cycle());
+    });
+  } else if (trimmer->should_trim_dirty() &&
+             cleaner->try_reserve_projected_usage(
+                 trimmer->get_trim_dirty_size_per_cycle())) {
     return trimmer->trim_dirty(
```

Comment:
Seems to me `trim_dirty()` and `trim_alloc()` are independent to each other. Probably we can run them concurrently.

This seems to be benefitial to both performance and simplicity, as trimmer can expose simpler interfaces to EPM, e.g. `should_trim()`, `trim_journal()` and `get_trim_size_per_cycle()`.

Does it also make sense to you @xxhdx1985126 ?

Reply:
> Seems to me `trim_dirty()` and `trim_alloc()` are independent to each other. Probably we can run them concurrently.
> 
> This seems to be benefitial to both performance and simplicity, as trimmer can expose simpler interfaces to EPM, e.g. `should_trim()`, `trim_journal()` and `get_trim_size_per_cycle()`.
> 
> Does it also make sense to you @xxhdx1985126 ?

Yeah, actually we have long been considering parallelise `trim_dirty()` and `trim_alloc()`.

The only issue I can think of here is that this probably mean that `trim_alloc()` and `trim_dirty()` can conflict on backref extents. But, since the relationship between `trim_alloc()` and `trim_dirty()` is pretty similar to that between client io and `trim_dirty()`, the conflict rate shouldn't be very high. And what's more, honestly speaking, I think we should probably face this kind of conflicts head-on, instead of keep avoiding it. Besides, when cold tier is involved, cleaners on hot and cold tier can also conflict on the root of the backref tree.

So to sum up, I favour the idea that `trim_alloc()` and `trim_dirty()` run concurrently.

Reply:
Worth to compare performance (before and after) to understand the impacts from concurrent trimming.